### DISCUSSION TITLE
How can I stop the gif playing from first load?

### DISCUSSION THREAD

**Discussion by shawnthye** [coil-kt/coil#1000] [Q&A]
# How can I stop the gif playing from first load?

As my requirement, I didn't want the `Gif` auto playing itself, 

I set `repeatCount(0)` but it still playing at least once

I am using the `ImageView extension` to load the Gif from network with `Crossfade` effect. 

Is there a way I can get `ScaleDrawable` or `AnimatedImageDrawable` from `CrossfadeDrawable`?

I tried just make a CustomImageView, and override the `setImageDrawable()`, and manually calling `CrossfadeDrawble#stop()`, but its doesn't work, and the Crossfade effect is gone too 😭

**Comment by colinrtwhite** [comment]
Hmm you'll likely manually have to apply the crossfade in this case. I would set `crossfade(false)`, create a custom `Target` that doesn't call `drawable.start` and then manually wrap it in a `CrossfadeDrawable` inside `Listener.onSuccess`.

**Reply by shawnthye** [reply]
> Hmm you'll likely manually have to apply the crossfade in this case. I would set `crossfade(false)`, create a custom `Target` that doesn't call `drawable.start` and then manually wrap it in a `CrossfadeDrawable` inside `Listener.onSuccess`.

Thanks @colinrtwhite 

But the roadblock is still the  `CrossfadeDrawable`, because `CrossfadeDrawable` is calling the child GIF drawable to start together?

```
class CrossfadeDrawable(
    start: Drawable?,
    end: Drawable?,
    val scale: Scale = Scale.FIT,
    val durationMillis: Int = DEFAULT_DURATION,
    val fadeStart: Boolean = true,
    val preferExactIntrinsicSize: Boolean = false
) : Drawable(), Drawable.Callback, Animatable2Compat {

    override fun start() {
        (start as? Animatable)?.start()
        (end as? Animatable)?.start()

        if (state != STATE_START) {
            return
        }

        state = STATE_RUNNING
        startTimeMillis = SystemClock.uptimeMillis()
        callbacks.forEachIndices { it.onAnimationStart(this) }

        invalidateSelf()
    }
}
```

**Reply by shawnthye** [reply]
else I could think of a custom transition like below

```
@OptIn(ExperimentalCoilApi::class)
class WrappedCrossFadeTransition(
    private val delegate: CrossfadeTransition = CrossfadeTransition(durationMillis = 2000),
) : Transition {
    override suspend fun transition(target: TransitionTarget, result: ImageResult) {

        if (result is SuccessResult) {
            val wrapped = CustomDrawable(result.drawable)
            delegate.transition(target, SuccessResult(wrapped, result.request, result.metadata))
        } else {
            delegate.transition(target, result)
        }
    }
}

class CustomDrawable(private val animatable: Drawable) : DrawableWrapper(animatable) {

    // sample, not fully implement
    fun start() {
        when (animatable) {
            is ScaleDrawable -> animatable.start()
            is AnimatedImageDrawable -> animatable.start()
        }
    }
}
```

but another question here, I wanna to play the Gif when user Touched and Hold on the GIF

How can I get back the `CustomDrawable` from **CrossfadeDrawable** ?  possible to change `start` and `end` drawable to be public accessible?

**Reply by colinrtwhite** [reply]
@shawnthye I think you're right that this isn't easily possible with the current API. `start` and `end` are private in 1.x due to bitmap pooling limitations, however I've made them public for the next 2.x alpha release here: https://github.com/coil-kt/coil/pull/1008

**Reply by shawnthye** [reply]
nice!! Thanks @colinrtwhite

**Reply by kasem-sm** [reply]
@shawnthye any updates/ guidelines on how to achieve this? Thanks!
